core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushPredicateIntoTableScan.java

```diff
@@ -185,6 +191,14 @@ private boolean arePlansSame(FilterNode filter, TableScanNode tableScan, PlanNod
                 .transformKeys(node.getAssignments()::get)
                 .intersect(node.getEnforcedConstraint());
 
+        Map<NodeRef<Expression>, Type> remainingExpressionTypes = typeAnalyzer.getTypes(session, symbolAllocator.getTypes(), decomposedPredicate.getRemainingExpression());
+        Optional<ConnectorExpression> translatedExpression = new ConnectorExpressionTranslator.SqlToConnectorExpressionTranslator(session, remainingExpressionTypes, metadata)
+                .process(decomposedPredicate.getRemainingExpression());
+        ConnectorExpression connectorExpression = translatedExpression.orElse(TRUE);
+        Expression remainingExpression = translatedExpression
+                .map(ignored -> (Expression) TRUE_LITERAL)
+                .orElse(decomposedPredicate.getRemainingExpression());
+
```

Comment:
The logic in this method turned somewhat convoluted and hard to follow. At a high level, what we should be doing is:

* Translate `decomposedPredicate.getRemainingExpression()` to a ConnectorExpression
* If the translation was successful, use the ConnectorExpression for pushdown
* If the "remaining expression" in the result of pushdown is Optional.empty(), use the original `decomposedPredicate.getRemainingExpression()` to assemble the result of this method
* Otherwise, if the "remaining expression"  in the result of pushdown is NOT Optional.empty(), translate it back to an Expression an use it to assemble the result of this method.

Reply:
Done

Reply:
I’ve opened issue #9506. I’m planning to implement it after this PR is merged.